timesketch/models/acl.py

```diff
@@ -0,0 +1,182 @@
+make it easy to annotate models to give them access to the ACL system.
+
+The model has the following permissionsL: "read", "write" and "delete".
+"""
+
+from flask_login import current_user
+from sqlalchemy import Column
+from sqlalchemy import ForeignKey
+from sqlalchemy import Integer
+from sqlalchemy import or_
+from sqlalchemy import not_
+from sqlalchemy import String
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.orm import relationship
+
+from timesketch.models import BaseModel
+from timesketch.models import db_session
+
+
+class AccessControlEntry(object):
+    """
+    Access Control Entry database model. It has a user object (instance of
+    timesketch.models.user.User) and a permission (read, write or delete).
+    """
+    @declared_attr
+    def user_id(self):
+        """Foreign key to a user model.
+
+        Returns:
+            A column (instance of sqlalchemy.Column)
+        """
+        return Column(Integer, ForeignKey('user.id'))
+
+    @declared_attr
+    def user(self):
+        """A relationship to a user object.
+
+        Returns:
+            A relationship (instance of sqlalchemy.orm.relationship)
+        """
+        return relationship('User')
+
+    # Permission column (read, write or delete)
+    permission = Column(String(255))
+
+
+class AccessControlMixin(object):
+    """
+    A MixIn for generating the necessary tables in the database and to make
+    it accessible from the parent model object (the model object that uses this
+    MixIn, i.e. the object that the ACL is added to).
+    """
+    @declared_attr
+    def acl(self):
+        """
+        Generates the ACL tables and adds the attribute to the parent model
+        object.
+
+        Returns:
+            A relationship to an ACE (timesketch.models.acl.AccessControlEntry)
+        """
+        self.AccessControlEntry = type(
+            '%sAccessControlEntry' % self.__name__,
+            (AccessControlEntry, BaseModel,),
+            dict(
+                __tablename__='%s_accesscontrolentry' % self.__tablename__,
+                parent_id=Column(
+                    Integer, ForeignKey('%s.id' % self.__tablename__)),
+                parent=relationship(self),
+            )
+        )
+        return relationship(self.AccessControlEntry)
+
+    @classmethod
+    def all_with_acl(cls, user=None):
+        """
+        Get all instances of the parent class that the user has read
+        permission on. I.e enforce ACL permission check when fetching from
+        the database.
+
+        Returns:
+            An ACL base query (instance of timesketch.models.AclBaseQuery)
+        """
+        # If no user, assume the user that made the request.
+        if not user:
+            user = current_user
+
+        return cls.query.filter(
+            or_(
+                cls.AccessControlEntry.user == user,
+                cls.AccessControlEntry.user == None),
```

Comment:
Yeah, I really want to, but in this case it is not possible. The way this query (SQLalchemy _or() filter) works is like this, and it throughs errors if not done this way.